sql-native/src/main/java/org/ballerinalang/sql/parameterprocessor/DefaultResultParameterProcessor.java

```diff
@@ -0,0 +1,761 @@
+import io.ballerina.runtime.api.values.BArray;
+import io.ballerina.runtime.api.values.BMap;
+import io.ballerina.runtime.api.values.BObject;
+import io.ballerina.runtime.api.values.BString;
+import io.ballerina.runtime.api.values.BXml;
+import org.ballerinalang.sql.Constants;
+import org.ballerinalang.sql.exception.ApplicationError;
+import org.ballerinalang.sql.utils.ColumnDefinition;
+import org.ballerinalang.sql.utils.ErrorGenerator;
+import org.ballerinalang.sql.utils.ModuleUtils;
+import org.ballerinalang.sql.utils.Utils;
+
+import java.math.BigDecimal;
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.CallableStatement;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.SQLXML;
+import java.sql.Statement;
+import java.sql.Struct;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.time.OffsetDateTime;
+import java.time.OffsetTime;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.List;
+import java.util.TimeZone;
+
+import static io.ballerina.runtime.api.utils.StringUtils.fromString;
+
+/**
+ * This class implements methods required convert SQL types into ballerina types and
+ * other methods that process the parameters of the result.
+ *
+ * @since 0.5.6
+ */
+public class DefaultResultParameterProcessor extends AbstractResultParameterProcessor {
+    private static final Object lock = new Object();
+    private static volatile DefaultResultParameterProcessor instance;
+
+    private static final ArrayType stringArrayType = TypeCreator.createArrayType(PredefinedTypes.TYPE_STRING);
+    private static final ArrayType booleanArrayType = TypeCreator.createArrayType(PredefinedTypes.TYPE_BOOLEAN);
+    private static final ArrayType intArrayType = TypeCreator.createArrayType(PredefinedTypes.TYPE_INT);
+    private static final ArrayType floatArrayType = TypeCreator.createArrayType(PredefinedTypes.TYPE_FLOAT);
+    private static final ArrayType decimalArrayType = TypeCreator.createArrayType(PredefinedTypes.TYPE_DECIMAL);
+
+    private static final Calendar calendar = Calendar
+            .getInstance(TimeZone.getTimeZone(Constants.TIMEZONE_UTC.getValue()));
+
+
+    public static DefaultResultParameterProcessor getInstance() {
+        if (instance == null) {
+            synchronized (lock) {
+                if (instance == null) {
+                    instance = new DefaultResultParameterProcessor();
+                }
+            }
+        }
+        return instance;
+    }
+
+    protected BArray createAndPopulateBBRefValueArray(Object firstNonNullElement, Object[] dataArray,
+                                                      Type type) throws ApplicationError {
+        BArray refValueArray = null;
+        int length = dataArray.length;
+        if (firstNonNullElement instanceof String) {
+            refValueArray = createEmptyBBRefValueArray(PredefinedTypes.TYPE_STRING);
+            for (int i = 0; i < length; i++) {
```

Comment:
It seems other than `BigDecimal` and `null` instances, this for loop is duplicated for every other instance. Maybe we can improve this by only initializing `refValueArray` and having a single for-loop before `return refValueArray;` (Line 144)

Reply:
We will change as suggested

Reply:
Resolved with 0bd8dba16b48436cd248d7e02ae4cc046b33a2ff